src/poll.rs

```diff
@@ -1,32 +1,34 @@
-use std::task::Context;
-
-use futures::{ready, task::Poll};
-use mio::Ready;
-use tokio::io::PollEvented;
+use futures::ready;
 
 use crate::error::TmqError::InterruptedSend;
 use crate::socket::AsZmqSocket;
-use crate::{socket::SocketWrapper, Multipart, Result};
-use std::{collections::VecDeque, ops::Deref};
+use crate::{Multipart, Result};
+use std::{
+    collections::VecDeque,
+    os::unix::io::RawFd,
+    task::{Context, Poll},
+};
+use tokio::io::unix::AsyncFd;
 use zmq::Socket;
 
 /// Implements functions for asynchronous reading and writing of multipart messages.
-///
-/// Uses a wrapped ZMQ socket. It needs to use a distinct inner type because
-/// [`tokio::io::PollEvented`] requires a [`mio::Evented`] parameter.
-pub(crate) struct ZmqPoller(PollEvented<SocketWrapper>);
+pub(crate) struct ZmqPoller {
+    socket: zmq::Socket,
+    async_fd: AsyncFd<RawFd>,
```

Comment:
This doesn't meet the API requirements of AsyncFd.  From [the tokio docs](https://docs.rs/tokio/1.0.1/tokio/io/unix/struct.AsyncFd.html):

> The AsyncFd takes ownership of an arbitrary object to represent the IO object. It is intended that this object will handle closing the file descriptor when it is dropped, avoiding resource leaks and ensuring that the AsyncFd can clean up the registration before closing the file descriptor. 

In your arrangement of the struct above, the socket might be dropped before the RawFd, and the AsyncFd then is holding an invalid file descriptor.

I suggest that ZmqPoller be instead defined as
```rust
pub(crate) struct ZmqPoller(AsyncFd<SocketWrapper>)
```

This requires changing SocketWrapper a bit.  You would need to do something like this in socket.rs:

```rust

/// Wrapper on top of a ZMQ socket.
///
/// The socket needs to be wrapped to allow various trait implementations.
pub(crate) struct SocketWrapper {
    pub(crate) socket: zmq::Socket,
    // This RawFd is held separately because it must be accessible 
    // without error after SocketWrapper initialization, for the AsRawFd trait.
    fd: RawFd,
}

impl SocketWrapper {
    pub fn new(socket: zmq::Socket) -> Result<Self> {
        Ok(Self { fd: socket.get_fd()?, socket })
    }
}

impl AsRawFd for SocketWrapper {
    fn as_raw_fd(&self) -> RawFd {
        self.fd
    }
}
```

Reply:
Right! I applied your suggestion. Thank you!